README.md
# WICE.js

WICE.js provides functions to interact with objects inside HTML canvas element using relative rectangle coordinates.
It is possible to design interactive exercises for educational purposes, captcha security verification, etc.

## Community

WICE.js is open source software released under an
[MIT license](https://github.com/yulioaj290/wice.js/blob/master/LICENSE).

You are welcome to
[report bugs](https://github.com/yulioaj290/wice.js/issues) or create pull
requests on [github](https://github.com/yulioaj290/wice.js).

## Installation

The easiest way to install acorn is with [`npm`][npm].

[npm]: https://www.npmjs.com/

```sh
npm install wice.js
```

Alternately, you can download the source and build acorn yourself:

```sh
git clone https://github.com/yulioaj290/wice.js.git
cd wice.js
npm install
npm run build
```

You must to include `wice.js` or `wice.min.js` file to your HTML page to get ready to use.

## Types of canvas exercises supported

The following are the main type of exercises that WICE.js support, but you may find some another type possible to perform with exiting functions.

* Order Exercise
* Position Object Exercise
* Select Point Exercise
* Associate Exercise
* Gap Match Exercise

## General properties

Property | Value | Description
-------- | ----- | -----------
canvasElementId | String <br /> Eg.: "myCanvas" | ID attribute of the canvas HTML element to link with
canvasWidth | Integer <br /> Eg.: 500 | Width of the canvas element (1)
imageSrc | String <br /> Eg.: "./my-image.jpg" | Image path to load on the canvas element
lineWidth | Integer <br /> Eg.: 2 | Width used when draw lines over canvas element (1)
lineColor | String <br /> Eg.: "#88FFCC" | Color used when draw lines (2)
fontFamily | String <br /> Eg.: "Arial" | Type of font used when draw texts over canvas element
fontSize | Integer <br /> Eg.: 24 | Font size for text over canvas (1)
fontColor | String <br /> Eg.: "#BB9922" | Font color for text over canvas (2)
fontStyle | String <br /> Eg.: "italic" | Font style, relative to `normal`, `bold`, `italic`, `small-caps`
fontAlign | String <br /> Eg.: "center" | Font alignment, relative to `center`, `left`, `right`, `start`, `end`
fontBaseline | String <br /> Eg.: "bottom" | Text baseline, relative to `middle`, `top`, `bottom`, `Alphabetic`, `Hanging`
mark | String <br /> Eg.: "x" | Define text mark to use to design a position over canvas. It can be a letter or a single word
orderType | Constant <br /> `ORDER_TYPE_ALPHA | ORDER_TYPE_NUM` | Alphabetic or numeric ordering (3)

<sup>1</sup> Sizes are in pixels (px)

<sup>2</sup> Colors can be written using Hexadecimal (`#FF88CC`), RGB (`rgb(100, 255, 190)`) or simple word (`blue, green, red`) format.

<sup>3</sup> Constants defined in CanvasExercise class.


#### Changing property values

```javascript
// Create canvas exercice (Associate)
const ex = new AssociateCanvasExercise('myCanvas');

// Setting canvas element width
ex.canvasWidth = 700;

// Assign the path to the image
ex.imageSrc = 'img/my-image.jpg';

// initializing canvas
ex.initializeCanvas();
```


## Order Exercise

Allow to perform order assignment to objects over canvas using coordinates

#### Properties

Property | Value | Description
-------- | ----- | -----------
listObjectsCoords | Array <br /> Eg.: `[[50,29,86,71],[92, 102, 156, 148]]` | Coordinates of the objects over canvas element
strokeRectObject | Boolean <br /> Eg.: true | Define if object will be highlighted using a rectangle based on the object coordinates

#### Code sample

```javascript
// Define objects coordinates over canvas element
var coords = [[80, 35, 375, 283], [395, 120, 715, 370], [36, 352, 88, 404]];

// Create canvas exercice (Order)
const ex = new OrderCanvasExercise('myCanvas');

// Setting coordinates property
ex.listObjectsCoords = coords;

// Setting property to highlight object with rectangle
ex.strokeRectObject = true;

// initializing canvas
ex.initializeCanvas();
```


## Position Object & Select Point Exercises

Allow to draw a mark in the position over canvas to identify objects using coordinates

#### Properties

Property | Value | Description
-------- | ----- | -----------
listObjectsCoords | Array <br /> Eg.: `[[50,29,86,71],[92, 102, 156, 148]]` | Coordinates of the objects over canvas element
strokeRectObject | Boolean <br /> Eg.: true | Define if object will be highlighted using a rectangle based on the object coordinates

#### Code sample

```javascript
// Define objects coordinates over canvas element
// You can use one of the following definitions
var coords = [];

// For several objects (Position Object)
coords = [[80, 35, 375, 283], [395, 120, 715, 370], [36, 352, 88, 404]];

// For one object (Select Point)
coords = [[80, 35, 375, 283]];

// Create canvas exercice (Position Object)
const ex = new PositionObjectCanvasExercise('myCanvas');

// Setting coordinates property
ex.listObjectsCoords = coords;

// Setting property to highlight object with rectangle
ex.strokeRectObject = true;

// initializing canvas
ex.initializeCanvas();
```


## Associate Exercises

Allow to draw a line between two points to associate two objects over canvas, identified using coordinates

#### Properties

Property | Value | Description
-------- | ----- | -----------
listObjectsCoords | Array <br /> Eg.: `[[50,29,86,71],[92, 102, 156, 148]]` | Coordinates of the objects over canvas element
listObjectsConnections | Array <br /> Eg.: `[[[92, 102, 156, 148]], [[50,29,86,71]]]` | Coordinates of connections between objects over canvas element
strokeRectObject | Boolean <br /> Eg.: true | Define if object will be highlighted using a rectangle based on the object coordinates

#### Code sample

```javascript
// Define objects coordinates over canvas element
var coords = [[80, 35, 375, 283], [395, 120, 715, 370], [36, 352, 88, 404]];
var connections = [[[36, 352, 88, 404], [36, 352, 88, 404]], [], [[80, 35, 375, 283]]];

// Create canvas exercice (Associate)
const ex = new AssociateCanvasExercise('myCanvas');

// Setting coordinates and associations properties
ex.listObjectsCoords = coords;
ex.listObjectsConnections = connections;

// Setting property to highlight object with rectangle
ex.strokeRectObject = true;

// initializing canvas
ex.initializeCanvas();
```

The examples of values in the previous example, establish a correspondence using array indexes, like following:

```
Objects Coordinates (3 Objects)
[[80, 35, 375, 283], [395, 120, 715, 370], [36, 352, 88, 404]]

[0] => [80, 35, 375, 283]   (Obj 1)
[1] => [395, 120, 715, 370] (Obj 2)
[2] => [36, 352, 88, 404]   (Obj 3)

Objects Connections
[[[36, 352, 88, 404], [395, 120, 715, 370]], [], [[80, 35, 375, 283]]]

[0] => [[36, 352, 88, 404], [395, 120, 715, 370]]
[1] => []
[2] => [[80, 35, 375, 283]]

Final Objects Connections
(Obj 1) -connected-to--> (Obj 3) & (Obj 2)
(Obj 2) -connected-to--> (none object)
(Obj 3) -connected-to--> (Obj 1)

```

## Gap Match Exercises

Allow to draw a text over objects inside the canvas element, identified using coordinates

#### Properties

Property | Value | Description
-------- | ----- | -----------
listObjectsCoords | Array <br /> Eg.: `[[50,29,86,71],[92, 102, 156, 148]]` | Coordinates of the objects over canvas element
listObjectsTags | Array <br /> Eg.: `['Human', 'Car']` | Tags associated to objects stored in `listObjectsCoords`
strokeRectObject | Boolean <br /> Eg.: true | Define if object will be highlighted using a rectangle based on the object coordinates

#### Code sample

```javascript
// Define objects coordinates over canvas element
var coords = [[80, 35, 375, 283], [395, 120, 715, 370], [36, 352, 88, 404]];
var tags = ['Human', 'Car', 'Building'];

// Create canvas exercice (Gap Match)
const ex = new GapMatchCanvasExercise('myCanvas');

// Setting coordinates and tags properties
ex.listObjectsCoords = coords;
ex.listObjectsConnections = tags;

// Setting property to highlight object with rectangle
ex.strokeRectObject = true;

// initializing canvas
ex.initializeCanvas();
```

The examples of values in the previous example, establish a correspondence using array indexes, like following:

```
Objects Coordinates (3 Objects)
[[80, 35, 375, 283], [395, 120, 715, 370], [36, 352, 88, 404]]

Objects Tags
['Human', 'Car', 'Building']

[0] => [80, 35, 375, 283]   (Human)
[1] => [395, 120, 715, 370] (Car)
[2] => [36, 352, 88, 404]   (Building)
```



CHANGELOG.md
## 0.0.3

 - Fix typos in README.md
 - Update home page of Docs website

## 0.0.2

 - Fix bug in gulpfile.js
 - Re-write

## 0.0.1

 - initial version

AssociateCanvasExercise-compiled.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Project: wice.js.
 * Author: Yulio Aleman Jimenez [@yulioaj290]
 * Date: 22/09/18
 * Created by WebStorm
 */

/**
 * AssociateCanvasExercise extends from CanvasExercise superclass.
 * Define the associate object exercises
 */
var AssociateCanvasExercise = function (_CanvasExercise) {
    _inherits(AssociateCanvasExercise, _CanvasExercise);

    /**
     * Constructor of the AssociateCanvasExercise class.
     * @param canvasElementId
     * @param canvasWidth
     * @param imageSrc
     * @param lineWidth
     * @param lineColor
     * @param lineHead
     * @param fontFamily
     * @param fontSize
     * @param fontColor
     * @param fontStyle
     * @param fontAlign
     * @param fontBaseline
     * @param mark
     * @param orderType
     * @param listObjectsCoords
     * @param listObjectsConnections
     * @param strokeRectObject
     */
    function AssociateCanvasExercise(canvasElementId) {
        var canvasWidth = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';
        var imageSrc = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : '';
        var lineWidth = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : '2';
        var lineColor = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : '#000000';
        var lineHead = arguments.length > 5 && arguments[5] !== undefined ? arguments[5] : false;
        var fontFamily = arguments.length > 6 && arguments[6] !== undefined ? arguments[6] : 'Arial';
        var fontSize = arguments.length > 7 && arguments[7] !== undefined ? arguments[7] : '20';
        var fontColor = arguments.length > 8 && arguments[8] !== undefined ? arguments[8] : '#000000';
        var fontStyle = arguments.length > 9 && arguments[9] !== undefined ? arguments[9] : 'regular';
        var fontAlign = arguments.length > 10 && arguments[10] !== undefined ? arguments[10] : 'center';
        var fontBaseline = arguments.length > 11 && arguments[11] !== undefined ? arguments[11] : 'middle';
        var mark = arguments.length > 12 && arguments[12] !== undefined ? arguments[12] : 'X';
        var orderType = arguments.length > 13 && arguments[13] !== undefined ? arguments[13] : ORDER_TYPE_NUM;
        var listObjectsCoords = arguments.length > 14 && arguments[14] !== undefined ? arguments[14] : [];
        var listObjectsConnections = arguments.length > 15 && arguments[15] !== undefined ? arguments[15] : [];
        var strokeRectObject = arguments.length > 16 && arguments[16] !== undefined ? arguments[16] : true;

        _classCallCheck(this, AssociateCanvasExercise);

        // own properties of the class
        var _this = _possibleConstructorReturn(this, (AssociateCanvasExercise.__proto__ || Object.getPrototypeOf(AssociateCanvasExercise)).call(this, canvasElementId, canvasWidth, imageSrc, lineWidth, lineColor, lineHead, fontFamily, fontSize, fontColor, fontStyle, fontAlign, fontBaseline, mark, orderType));

        _this._listObjectsCoords = listObjectsCoords; // array of objects coordinates in the image
        _this._listObjectsConnections = listObjectsConnections; // array of objects connections in the image
        _this._listObjectsAssociated = []; // array to store the array of objects coordinates associated by user
        _this._strokeRectObject = strokeRectObject; // define if object will be put into a rectangle
        _this._beginClickCoord = []; // store the coordinates of the 1st click event in format [x,y]
        _this._numberAssociationsConnected = 0;

        // binding click event to canvas element to allow the associate object exercise execution
        _this._canvasElement.addEventListener("click", _this.clickAction.bind(_this), false);
        return _this;
    }

    /**
     * Perform an associate action using a line between objects based on two click event over canvas element.
     * This function uses the coordinates of the objects inside canvas element and the coordinates of the click event.
     * @param event
     * @returns {*}
     */


    _createClass(AssociateCanvasExercise, [{
        key: 'clickAction',
        value: function clickAction(event) {

            // Obtain mouse click position
            var current_x = event.pageX - event.currentTarget.offsetLeft,
                current_y = event.pageY - event.currentTarget.offsetTop;
            // alert(`cursorX: ${cur_x}, cursorY: ${cur_y}`);

            if (this._canvasElement.getContext) {
                // getting 2d context from canvas element
                var ctx = this._canvasElement.getContext("2d");

                // asking if there is Associations to be perform
                if (this._numberAssociationsConnected < this._getNumberOfConnections()) {

                    // asking if it's the 1st click event
                    if (this._beginClickCoord.length === 0) {
                        // Storing the 1st click event coordinates
                        this._beginClickCoord = [current_x, current_y];
                    } else {

                        // getting coordinates of real objects, begin & end objects
                        var beginObject = this._findObjectThroughCoords(this._beginClickCoord),
                            endObject = this._findObjectThroughCoords([current_x, current_y]),
                            connection = this._findConnection(beginObject, endObject);

                        if (!!beginObject && !!endObject) {
                            if (connection === 1) {
                                this._listObjectsAssociated.push([beginObject, endObject]);
                            } else if (connection === 0) {
                                this._listObjectsAssociated.push([endObject, beginObject]);
                            }
                        }

                        ctx.lineWidth = this._lineWidth;
                        ctx.strokeStyle = this._lineColor;
                        ctx.beginPath();
                        ctx.moveTo(this._beginClickCoord[0], this._beginClickCoord[1]);
                        ctx.lineTo(current_x, current_y);
                        ctx.stroke();

                        this._numberAssociationsConnected++;

                        // Resetting begin click coordinates
                        this._beginClickCoord = [];
                    }

                    return this.getListObjectsAssociated();
                }
            }
        }
    }, {
        key: '_findConnection',
        value: function _findConnection(beginObject, endObject) {
            if (this._findEndCoordsInConnections(endObject, this._findBeginCoordsInConnections(beginObject))) {
                return 1;
            } else if (this._findEndCoordsInConnections(beginObject, this._findBeginCoordsInConnections(endObject))) {
                return 0;
            } else {
                return -1;
            }
        }
    }, {
        key: '_findBeginCoordsInConnections',
        value: function _findBeginCoordsInConnections(beginObject) {
            // looping the list of objects coordinates
            for (var i = 0; i < this._listObjectsCoords.length; i++) {

                var value1 = JSON.stringify(this._listObjectsCoords[i]),
                    value2 = JSON.stringify(beginObject);

                // looking for the index of the begin object
                if (value1 === value2) {
                    return i;
                }
            }
        }
    }, {
        key: '_findEndCoordsInConnections',
        value: function _findEndCoordsInConnections(endObject, index) {
            // looping the list of objects connections coordinates
            for (var i = 0; i < this._listObjectsConnections[index].length; i++) {

                var value1 = JSON.stringify(this._listObjectsConnections[index][i]),
                    value2 = JSON.stringify(endObject);

                // looking for the index of the begin object
                if (value1 === value2) {
                    return true;
                }
            }

            return false;
        }
    }, {
        key: '_findObjectThroughCoords',
        value: function _findObjectThroughCoords(coords) {

            // declaring variables
            var x = void 0,
                y = void 0,
                x1 = void 0,
                y1 = void 0;

            // looping the list of objects coordinates
            for (var i = 0; i < this._listObjectsCoords.length; i++) {

                // Coordinates of the images
                // The canvas divisor allow to calc the exact position of every coordinate,
                // even if the canvas element is positioned anywhere inside the web page
                x = Math.round(this._listObjectsCoords[i][0] / this._canvasDivisor);
                y = Math.round(this._listObjectsCoords[i][1] / this._canvasDivisor);
                x1 = Math.round(this._listObjectsCoords[i][2] / this._canvasDivisor);
                y1 = Math.round(this._listObjectsCoords[i][3] / this._canvasDivisor);

                // asking if a valid object was clicked, based on coordinates of objects and click event
                if (x <= coords[0] && coords[0] <= x1 && y <= coords[1] && coords[1] <= y1) {

                    return this._listObjectsCoords[i];
                }
            }

            return false;
        }
    }, {
        key: '_getNumberOfConnections',
        value: function _getNumberOfConnections() {
            var counter = 0;
            for (var i = 0; i < this._listObjectsConnections.length; i++) {
                counter += this._listObjectsConnections[i].length;
            }
            return counter;
        }

        /**
         * Return and print in JS console the array of the list of associated objects
         * @returns {Array}
         */

    }, {
        key: 'getListObjectsAssociated',
        value: function getListObjectsAssociated() {
            console.log(JSON.stringify(this._listObjectsAssociated));
            return this._listObjectsAssociated;
        }
    }]);

    return AssociateCanvasExercise;
}(CanvasExercise);

AssociateCanvasExercise.exports = AssociateCanvasExercise;

// Live execution of code
var coords = [[80, 35, 375, 283], [395, 120, 715, 370], [36, 352, 88, 404], [520, 20, 590, 76]],
    connections = [[[36, 352, 88, 404], [520, 20, 590, 76]], [[36, 352, 88, 404]], [[395, 120, 715, 370], [520, 20, 590, 76]], [[80, 35, 375, 283], [395, 120, 715, 370]]];

var ex = new AssociateCanvasExercise('myCanvas');
ex._listObjectsCoords = coords;
ex._listObjectsConnections = connections;
ex._canvasWidth = 700;
ex._imageSrc = 'plasma-desktop.jpg';
ex.initializeCanvas();

var marks = [];

for (var i = 0; i < ex._getNumberOfConnections(); i++) {
    marks[i] = " | ";
}

document.getElementById("options").innerText = JSON.stringify(marks);

ex._canvasElement.addEventListener("click", function () {
    if (ex._beginClickCoord.length === 0) {
        marks.shift();
        document.getElementById("options").innerText = JSON.stringify(marks);
        document.getElementById("result").innerText = JSON.stringify(ex.getListObjectsAssociated());
    }
}, false);

console.log(ex);

//# sourceMappingURL=AssociateCanvasExercise-compiled.js.map